clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2366,6 +2751,26 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
     return FD;
   }
 
+  static FunctionDecl *createFreeFunctionDecl(ASTContext &Ctx, FunctionDecl *FD,
+                                              SourceLocation Loc,
+                                              bool IsInline) {
+    // Create this with no prototype, and we can fix this up after we've seen
+    // all the params.
+    FunctionProtoType::ExtProtoInfo Info(CC_OpenCLKernel);
+    QualType FuncType = Ctx.getFunctionType(Ctx.VoidTy, {}, Info);
+    const IdentifierInfo *NewIdent =
+        &Ctx.Idents.get(constructFFKernelName(Ctx, FD));
```

Comment:
Maybe I am missing something obvious here but why are you calling this and not `constructFreeFunctionKernelName`

Reply:
At this point the name of the function is a placeholder. We have just created a FunctionDecl and no parameters have been added. We will be adding parameters one by one. Once all the Visitors have run, the FunctionDecl will have been fully created, with parameters and a body. Then, in Sema::SetSYCLKernelNames() we set the name by which the kernel will be recognized from the host. At this point we can extract the full mangled name. We couldn't see the mangled name earlier because the FunctionDecl had not been created fully with parameters.